components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java

```diff
@@ -1002,21 +1025,32 @@ private void addURITemplates(int apiId, API api, int tenantId) throws APIManagem
      */
     private void registerOrUpdateResourceInKeyManager(API api, String tenantDomain) throws APIManagementException {
         //get new key manager instance for  resource registration.
-        KeyManager keyManager = KeyManagerHolder.getKeyManagerInstance(tenantDomain);
-        Map registeredResource = keyManager.getResourceByApiId(api.getId().toString());
-        if (registeredResource == null) {
-            boolean isNewResourceRegistered = keyManager.registerNewResource(api, null);
-            if (!isNewResourceRegistered) {
-                handleException("APIResource registration is failed while adding the API- " + api.getId().getApiName()
-                        + "-" + api.getId().getVersion());
-            }
-        } else {
-            //update APIResource.
-            String resourceId = (String) registeredResource.get("resourceId");
-            if (resourceId == null) {
-                handleException("APIResource update is failed because of empty resourceID.");
+        Map<String, KeyManagerDto> tenantKeyManagers = KeyManagerHolder.getTenantKeyManagers(tenantDomain);
+        for (Map.Entry<String, KeyManagerDto> keyManagerDtoEntry : tenantKeyManagers.entrySet()) {
+            KeyManager keyManager = keyManagerDtoEntry.getValue().getKeyManager();
+            if (keyManager != null) {
+                try {
+                    Map registeredResource = keyManager.getResourceByApiId(api.getId().toString());
+                    if (registeredResource == null) {
+                        boolean isNewResourceRegistered = keyManager.registerNewResource(api, null);
+                        if (!isNewResourceRegistered) {
+                            log.warn("APIResource registration is failed while adding the API- " +
+                                    api.getId().getApiName()
+                                    + "-" + api.getId().getVersion() + " into KeyManager : " +
+                                    keyManagerDtoEntry.getKey());
+                        }
+                    } else {
+                        //update APIResource.
+                        String resourceId = (String) registeredResource.get("resourceId");
+                        if (resourceId == null) {
+                            handleException("APIResource update is failed because of empty resourceID.");
+                        }
+                        keyManager.updateRegisteredResource(api, registeredResource);
+                    }
+                } catch (APIManagementException e) {
+                    log.error("API Resource Registration failed in KeyManager " + keyManagerDtoEntry.getKey(), e);
```

Comment:
Please change all occurrences of "KeyManager" in logs and messages to "Key Manager".